File: backend/src/utils/file_utils.py
"""File utility functions."""

import hashlib
import os
from typing import Optional

from ..models import FileType


# Binary file signatures (magic bytes)
BINARY_SIGNATURES = [
    b"\x89PNG",  # PNG
    b"\xff\xd8\xff",  # JPEG
    b"GIF87a",  # GIF
    b"GIF89a",  # GIF
    b"PK\x03\x04",  # ZIP
    b"PK\x05\x06",  # ZIP (empty)
    b"\x1f\x8b",  # GZIP
    b"BZ",  # BZIP2
    b"\x7fELF",  # ELF executable
    b"MZ",  # DOS/Windows executable
    b"\x00\x00\x01\x00",  # ICO
    b"\x00\x00\x02\x00",  # CUR
    b"RIFF",  # RIFF (WAV, AVI)
    b"\x00\x00\x00",  # Various binary formats
]

# Text file extensions
TEXT_EXTENSIONS = {
    ".md",
    ".markdown",
    ".txt",
    ".text",
    ".yaml",
    ".yml",
    ".json",
    ".xml",
    ".html",
    ".htm",
    ".css",
    ".js",
    ".ts",
    ".jsx",
    ".tsx",
    ".py",
    ".rb",
    ".go",
    ".rs",
    ".java",
    ".c",
    ".cpp",
    ".h",
    ".hpp",
    ".sh",
    ".bash",
    ".zsh",
    ".fish",
    ".sql",
    ".graphql",
    ".toml",
    ".ini",
    ".cfg",
    ".conf",
    ".env",
    ".gitignore",
    ".dockerignore",
    ".editorconfig",
    ".prettierrc",
    ".eslintrc",
}


def is_binary_file(file_path: str, content: Optional[bytes] = None) -> bool:
    """Detect if a file is binary.

    Args:
        file_path: Path to the file.
        content: Optional file content bytes.

    Returns:
        True if file is binary, False if text.
    """
    # Check extension first
    ext = os.path.splitext(file_path)[1].lower()
    if ext in TEXT_EXTENSIONS:
        return False

    # If content provided, check magic bytes
    if content:
        for sig in BINARY_SIGNATURES:
            if content.startswith(sig):
                return True

        # Check for null bytes (common in binary files)
        if b"\x00" in content[:1024]:
            return True

    # If no content and unknown extension, try to read
    if content is None and os.path.exists(file_path):
        try:
            with open(file_path, "rb") as f:
                sample = f.read(1024)
                return is_binary_file(file_path, sample)
        except Exception:
            return True

    return False


def get_file_type(file_path: str, is_directory: bool = False) -> FileType:
    """Determine the file type from path.

    Args:
        file_path: Path to the file.
        is_directory: Whether the path is a directory.

    Returns:
        FileType enum value.
    """
    if is_directory:
        return FileType.DIRECTORY

    ext = os.path.splitext(file_path)[1].lower()

    if ext in {".md", ".markdown"}:
        return FileType.MARKDOWN
    elif ext in {".yaml", ".yml"}:
        return FileType.YAML
    elif ext in TEXT_EXTENSIONS:
        return FileType.TEXT
    else:
        # Check if binary
        if os.path.exists(file_path):
            if is_binary_file(file_path):
                return FileType.BINARY
        return FileType.TEXT


def compute_file_hash(content: bytes) -> str:
    """Compute MD5 hash of file content.

    Args:
        content: File content bytes.

    Returns:
        Hexadecimal MD5 hash string.
    """
    return hashlib.md5(content).hexdigest()


def get_file_size(file_path: str) -> int:
    """Get file size in bytes.

    Args:
        file_path: Path to the file.

    Returns:
        File size in bytes.
    """
    return os.path.getsize(file_path)


def read_text_file(file_path: str, encoding: str = "utf-8") -> Optional[str]:
    """Read text file content.

    Args:
        file_path: Path to the file.
        encoding: Text encoding.

    Returns:
        File content as string, or None if binary.
    """
    if is_binary_file(file_path):
        return None

    try:
        with open(file_path, "r", encoding=encoding) as f:
            return f.read()
    except UnicodeDecodeError:
        return None
    except Exception:
        return None

File: packages/backend/src/utils/file-utils.ts
/**
 * File utility functions.
 */

import { createHash } from 'crypto';
import { readFile, stat } from 'fs/promises';
import { extname } from 'path';

import type { FileType } from '@skills-runtime/shared';

// Binary file signatures (magic bytes)
const BINARY_SIGNATURES: Uint8Array[] = [
  new Uint8Array([0x89, 0x50, 0x4e, 0x47]), // PNG
  new Uint8Array([0xff, 0xd8, 0xff]), // JPEG
  new Uint8Array([0x47, 0x49, 0x46, 0x38, 0x37, 0x61]), // GIF87a
  new Uint8Array([0x47, 0x49, 0x46, 0x38, 0x39, 0x61]), // GIF89a
  new Uint8Array([0x50, 0x4b, 0x03, 0x04]), // ZIP
  new Uint8Array([0x50, 0x4b, 0x05, 0x06]), // ZIP (empty)
  new Uint8Array([0x1f, 0x8b]), // GZIP
  new Uint8Array([0x42, 0x5a]), // BZIP2
  new Uint8Array([0x7f, 0x45, 0x4c, 0x46]), // ELF executable
  new Uint8Array([0x4d, 0x5a]), // DOS/Windows executable
  new Uint8Array([0x00, 0x00, 0x01, 0x00]), // ICO
  new Uint8Array([0x00, 0x00, 0x02, 0x00]), // CUR
  new Uint8Array([0x52, 0x49, 0x46, 0x46]), // RIFF (WAV, AVI)
];

// Text file extensions
const TEXT_EXTENSIONS = new Set([
  '.md',
  '.markdown',
  '.txt',
  '.text',
  '.yaml',
  '.yml',
  '.json',
  '.xml',
  '.html',
  '.htm',
  '.css',
  '.js',
  '.ts',
  '.jsx',
  '.tsx',
  '.py',
  '.rb',
  '.go',
  '.rs',
  '.java',
  '.c',
  '.cpp',
  '.h',
  '.hpp',
  '.sh',
  '.bash',
  '.zsh',
  '.fish',
  '.sql',
  '.graphql',
  '.toml',
  '.ini',
  '.cfg',
  '.conf',
  '.env',
  '.gitignore',
  '.dockerignore',
  '.editorconfig',
  '.prettierrc',
  '.eslintrc',
]);

/**
 * Check if content starts with a signature.
 */
function startsWithSignature(content: Uint8Array, signature: Uint8Array): boolean {
  if (content.length < signature.length) {
    return false;
  }
  for (let i = 0; i < signature.length; i++) {
    if (content[i] !== signature[i]) {
      return false;
    }
  }
  return true;
}

/**
 * Detect if content is binary.
 *
 * @param filePath - Path to the file.
 * @param content - Optional file content bytes.
 * @returns True if file is binary, false if text.
 */
export function isBinaryContent(filePath: string, content?: Uint8Array): boolean {
  // Check extension first
  const ext = extname(filePath).toLowerCase();
  if (TEXT_EXTENSIONS.has(ext)) {
    return false;
  }

  // If content provided, check magic bytes
  if (content) {
    for (const sig of BINARY_SIGNATURES) {
      if (startsWithSignature(content, sig)) {
        return true;
      }
    }

    // Check for null bytes (common in binary files)
    const sampleLength = Math.min(1024, content.length);
    for (let i = 0; i < sampleLength; i++) {
      if (content[i] === 0) {
        return true;
      }
    }
  }

  return false;
}

/**
 * Detect if a file is binary by reading from disk.
 *
 * @param filePath - Path to the file.
 * @returns True if file is binary, false if text.
 */
export async function isBinaryFile(filePath: string): Promise<boolean> {
  // Check extension first
  const ext = extname(filePath).toLowerCase();
  if (TEXT_EXTENSIONS.has(ext)) {
    return false;
  }

  try {
    const buffer = await readFile(filePath);
    const sample = new Uint8Array(buffer.buffer, buffer.byteOffset, Math.min(1024, buffer.length));
    return isBinaryContent(filePath, sample);
  } catch {
    return true;
  }
}

/**
 * Determine the file type from path.
 *
 * @param filePath - Path to the file.
 * @param isDirectory - Whether the path is a directory.
 * @param content - Optional file content for binary detection.
 * @returns FileType value.
 */
export function getFileType(
  filePath: string,
  isDirectory: boolean = false,
  content?: Uint8Array
): FileType {
  if (isDirectory) {
    return 'directory';
  }

  const ext = extname(filePath).toLowerCase();

  if (ext === '.md' || ext === '.markdown') {
    return 'markdown';
  } else if (ext === '.yaml' || ext === '.yml') {
    return 'yaml';
  } else if (TEXT_EXTENSIONS.has(ext)) {
    return 'text';
  } else if (content && isBinaryContent(filePath, content)) {
    return 'binary';
  }

  return 'text';
}

/**
 * Compute MD5 hash of file content.
 *
 * @param content - File content bytes.
 * @returns Hexadecimal MD5 hash string.
 */
export function computeFileHash(content: Uint8Array | Buffer): string {
  return createHash('md5').update(content).digest('hex');
}

/**
 * Get file size in bytes.
 *
 * @param filePath - Path to the file.
 * @returns File size in bytes.
 */
export async function getFileSize(filePath: string): Promise<number> {
  const stats = await stat(filePath);
  return stats.size;
}

/**
 * Read text file content.
 *
 * @param filePath - Path to the file.
 * @param encoding - Text encoding.
 * @returns File content as string, or null if binary.
 */
export async function readTextFile(
  filePath: string,
  encoding: BufferEncoding = 'utf-8'
): Promise<string | null> {
  if (await isBinaryFile(filePath)) {
    return null;
  }

  try {
    const content = await readFile(filePath, { encoding });
    return content;
  } catch {
    return null;
  }
}
